### ISSUE TITLE
Enhancement: [no-unnecessary-condition] Option to only enable 'optional chain' related checks

### ISSUE THREAD

**Issue by reteps** [typescript-eslint/typescript-eslint#11518]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unnecessary-condition/#options

### Description

Our project gets a lot of false positives related to `Unnecessary conditional` due to incomplete types in our project. It would be great if we could enable _only_ the "Unnecessary optional chain" errors to aid us in incremental migration.

### Fail

```TypeScript
This should still flag as `Unnecessary optional chain on a non-nullish value.`


    const option = courseSelect.selectedOptions[0];

    if (option) {
      copyForm.action = option?.dataset.copyUrl ?? '';
```

### Pass

```TypeScript
This should still be allowed (previously: ``Unnecessary conditional, comparison is always true, since `"expire" === "expire"` is true.``):


    if (action === 'set_total') {
      explanation =
        'Updating the total time limit will set the given amount of time for the assessment based on when the assessment was started.';
    } else if (action === 'expire') {
      explanation =
        'This will expire the time limit and students will be unable to submit any further answers.';
    }
```

**Comment by bradzacher** [comment]
The pass case you have listed has been discussed previously and the decision was that we weren't going to change the rule because the rule is correctly reporting on that case.

If you are intentionally adding an unnecessary check for stylistic exhaustive reasons with the plan of adding more in future  - then a disable comment is one option.

The better option is to instead use a `switch` which can be enforced to be exhaustive via other lint rules and can also be easily enforced to be exhaustive via a `default` check against `never`.

Additionally you state

> to aid us in incremental migration 

We do not design lint rules for migrations or incremental adoptions. This is on purpose because options live forever and incur a large maintenance cost for us - but migrations are temporary.

If you are migrating we instead suggest options like using the warning severity and allowing warnings in CI.

So right now this is a "won't fix".

---

But before we commit to that and close I'd like to understand what you mean by

> due to incomplete types in our project

What are "incomplete types"?

**Comment by reteps** [comment]
The rationale shown for switch cases makes sense to me. I wanted to raise a few other common cases:

```ts
let anyCourseHadJsonErrors = false;
await async.eachOfSeries(Array.from(courseDirs), async (courseDir, index) => {
  const syncResult = await syncFromDisk.syncOrCreateDiskToSql(courseDir, job);
  if (syncResult.hadJsonErrors) anyCourseHadJsonErrors = true;
});

// Unnecessary conditional, value is always falsy.
if (anyCourseHadJsonErrors) {
  throw new Error(
    'One or more courses had JSON files that contained errors and were unable to be synced',
  );
}
```

```ts
let output: string | null = null;
try {
   await csvConverter.fromStream(csvStream).subscribe(async (json, number) => {
      // conditionally set output
      output = condition ? null : 'foo';
   })
}  finally {
  // Unnecessary conditional, comparison is always false, since `null != null` is false.
  if (output != null) {
    job.verbose(output);
  }
}
```

```ts
// Unnecessary conditional, value is always truthy.
if (courseData.courseInstances[courseInstanceId]) {

}
```

The third case is interesting, and probably been discussed before. Given the size of our project, it isn't practical to enable https://www.typescriptlang.org/tsconfig/#noUncheckedIndexedAccess. 

The documentation states:

> Another workaround is to use array.at(0) (which is always possibly-undefined) to indicate array access that may be out-of-bounds. Otherwise, a disable comment will often make sense for these kinds of cases.

This works for arrays, but for `Record` types, no similar construct exists, and the DX of a disable comment for every object access isn't great. 

Does this mean that this rule isn't just suited for our project given those two constraints?

- We don't want a major refactor to switch to `noUncheckedIndexedAccess`, which in many cases, isn't helpful
- We don't want to deal with incorrectly-flagged "unnecessary conditional, value is always truthy" messages.

I was thinking that a system where you can disable some of these messages by ID would be incredibly useful, since the `unnecessary conditional, value is always truthy` doesn't work for us, and presumably the other two cases I mentioned would also require a refactor to help typescript understand that it could be falsy.

**Comment by bradzacher** [comment]
The first case is easily fixed by annotating the variable with `boolean`.

The second case would be a good use of a disable comment to explain and educate. It by itself is not reason enough to warrant opting out of *_ALL_* checks though.

The third case [should work out of the box](https://typescript-eslint.io/play#ts=5.9.2&fileType=.ts&code=DYUwLgBAhgTjUE8BcEDOYYEsB2BzA2gLoQC8ERA3AFCYBmEAFLPAvgAyECUEA3lRBAD0giADpxVAL5UqAYwD22dBBggFMACYoASmvmaAPOix4ANGgw5cAPlK9J1Oo1XqNo1PIC2IANIgE3HwCwmISkkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false) - but without a repro I cannot advise further on this one.

**Comment by reteps** [comment]
Repro: https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=DYUwLgBAhgTjUE8BcEDOYYEsB2BzA2gLoQC8ERA3AFCYBmEAFLPAvgAyECUEA3lRBAD0giADpxVAL5UqAYwD22dBBggFMACYoASmvmaAPOix4ANBHkAjAFZqwAPlK9J1Oo1XqNo2vPnc%2BAsJiEpJAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

any record where the value isn't nullable, e.g. `Record<string, object>`. We are still checking this because we don't want a major refactor to switch to noUncheckedIndexedAccess, which in many cases, isn't helpful.

**Comment by reteps** [comment]
The first case cannot be solved by the annotation (repro: https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=JYWwDg9gTgLgBAKjgQwM4tQTwHYGM4BmUEIcA5GjrmQNwBQdANgKbzLaYDCEArlKswASyACYApVBGwBRKMX4AuOACMIEFuzgBeQskYD6yAO7JgbLHgB0zZLgAWAeQIBlZlGDNUACgDaAXQAaDCo4LwB9IOBsEWYADwBKbQA%2BOABvOjg4YAJQqJjY7S0dAAZE9i5efiFRCSlZeXQdGCgeZnoAX3j6OmzQ8u4%2BAWFxSRk5aFRE9MyAehm4S0W6dqA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

The second case (an async function setting a value) seems pretty common: https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=DYUwLgBA9grmAOcBcEDOYBOBLAdgcwgB8IcZhgIBeEs4AbgChRIAjKKC6gCgEYrLqAbR4BdQQAYRASgaYAnhADeDCBACGAdzVZIOEBogAFDFAC2WVCC5cMIVBwBuIADQRbAKxABjMFKoA%2BJRVVaDhESG42Dj8AfhpyCBQAcgAzdiTGELc7RysYHAATEBTcEAKZVQBfGUqIEpw1cgVlVQB6VogAVRw9LztUNQwFLyhCnSxRxtcR03hBi1GIC3VgLTlUOsbLV1RcPogAA1IEgEJqY%2BADpY2UrZAAOmCsFIguWAQ4CDP44D9lYMqDEBDBAAA94FAMJBFJUgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

If I cast them with

```
let anyCourseHadJsonErrors = false as boolean;
```

or

```
let output = null as string | null;
```

I don't get a type error, but with the type annotation, it still errors.

**Comment by reteps** [comment]
@bradzacher Any ideas?

**Comment by bradzacher** [comment]
> The second case (an async function setting a value) seems pretty common

It's not super common, no. TS has _never_ supported this case well and if you use this style of coding you'll notice the weakness.

If you're using something simple like `boolean` things will "just work" because of luck -- the value narrows to `false` which means you can use it like a boolean. But if you use a nullish type then you'll see how this style falls over.

For example your code will not typecheck as you expect:

```ts
  if (output != null) {
    output.toString(); // Error: Property 'toString' does not exist on type 'never'.
//  ^? output: never
  }
  output.toString(); // Error: 'output' is possibly 'null'.
//^? output: null
```

So writing your code in this manner doesn't satisfy the rule, nor does it satisfy TS itself.

All of these cases of modifying an external closure will have the same weakness and so it's something you should avoid at all costs.

For this reason it is much better to write `async` code with all the statements included in the same closure as it ensures that TS can understand it; which in turn means that our rules can understand it.

The fix is simple though, the assertion on the initial value to avoid TS narrowing it. I don't use this pattern much because it's not a great pattern with TS so I'd forgotten that it needs to be an assertion to avoid the automated control-flow narrowing.

**Comment by reteps** [comment]
Any ideas for the first case?

**Comment by bradzacher** [comment]
> any record where the value isn't nullable, e.g. Record<string, object>

Nitpick it's not to do with nullability, it's whether the type can be "falsey". `object` is guaranteed to be a non-primitive, non-nullish value, so it's never "falsey". But for example if you have `Record<string, string>`, `Record<string, number>` or `Record<string, boolean>` then all of these will pass the rule.

The easiest workaround if you know you're not using nullish values is to use a `in` check instead:

```ts
const record: Record<string, object> = {};
if ('foo' in record) {
  record.foo.toString();
}
```

**Comment by reteps** [comment]
Thanks @bradzacher . You can close this out. Thanks for the helpful tips on best practices here.
